Cannot import a MongoDB data source (creating one works fine)

<!---
Please note the following potential times when an issue might be in Terraform core:

* [Configuration Language](https://www.terraform.io/docs/configuration/index.html) or resource ordering issues
* [State](https://www.terraform.io/docs/state/index.html) and [State Backend](https://www.terraform.io/docs/backends/index.html) issues
* [Provisioner](https://www.terraform.io/docs/provisioners/index.html) issues
* [Registry](https://registry.terraform.io/) issues
* Spans resources across multiple providers

If you are running into one of these scenarios, we recommend opening an issue in the [Terraform core repository](https://github.com/hashicorp/terraform/) instead.
--->

### Terraform CLI Version
```
Terraform v1.9.7
```

<!--- Please run `terraform -v` to show the Terraform core version and provider version(s). If you are not running the latest version of Terraform or the provider, please upgrade because your issue may have already been fixed. [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions). --->

### Affected Resource(s)

<!--- Please list the affected resources and data sources. --->

* metabase_database

### Terraform Configuration Files

<!--- Information about code formatting: https://help.github.com/articles/basic-writing-and-formatting-syntax/#quoting-code --->

Please include all Terraform configurations required to reproduce the bug. Bug reports without a functional reproduction may be closed without investigation.

```hcl
resource "metabase_database" "mongo" {
  name   = "my-mongo"
  engine = "mongo"

  details = jsonencode({
    host    = "mongo_host"
    dbname  = "mongo_dbname"
    user    = "mongo_user"
    use-srv = true
    ssl     = true
  })

  details_secure = jsonencode({
    pass = "mongo_pass"
  })
}
```

### Debug Output

<!---
If possible, please provide the debug output as this will make it easier to identify the issue.

To obtain the debug output, see the [Terraform documentation on debugging](https://www.terraform.io/docs/internals/debugging.html).
--->
<details>
<summary>Debug output</summary>

```

2025-01-28T18:08:54.579+0200 [INFO]  Terraform version: 1.9.7
2025-01-28T18:08:54.579+0200 [DEBUG] using github.com/hashicorp/go-tfe v1.58.0
2025-01-28T18:08:54.579+0200 [DEBUG] using github.com/hashicorp/hcl/v2 v2.20.0
2025-01-28T18:08:54.579+0200 [DEBUG] using github.com/hashicorp/terraform-svchost v0.1.1
2025-01-28T18:08:54.579+0200 [DEBUG] using github.com/zclconf/go-cty v1.14.4
2025-01-28T18:08:54.579+0200 [INFO]  Go runtime version: go1.22.7
2025-01-28T18:08:54.579+0200 [INFO]  CLI args: []string{"terraform", "import", "metabase_database.mongo", "6"}
2025-01-28T18:08:54.579+0200 [DEBUG] File doesn't exist, but doesn't need to. Ignoring.
2025-01-28T18:08:54.580+0200 [INFO]  CLI command args: []string{"import", "metabase_database.mongo", "6"}
2025-01-28T18:08:54.590+0200 [DEBUG] checking for provisioner in "."
2025-01-28T18:08:54.590+0200 [DEBUG] checking for provisioner in "/usr/local/bin"
2025-01-28T18:08:54.593+0200 [DEBUG] Building and walking import graph
2025-01-28T18:08:54.593+0200 [DEBUG] Not attaching any node states: overall state is nil
2025-01-28T18:08:54.593+0200 [DEBUG] No state, no orphan outputs
2025-01-28T18:08:54.593+0200 [DEBUG] ProviderTransformer: "metabase_database.mongo (expand)" (*terraform.nodeExpandPlannableResource) needs provider["registry.terraform.io/bnjns/metabase"]
2025-01-28T18:08:54.593+0200 [DEBUG] created provider logger: level=debug
2025-01-28T18:08:54.593+0200 [INFO]  provider: configuring client automatic mTLS
2025-01-28T18:08:54.599+0200 [DEBUG] provider: starting plugin: path=.terraform/providers/registry.terraform.io/bnjns/metabase/0.9.1/darwin_arm64/terraform-provider-metabase_v0.9.1 args=[".terraform/providers/registry.terraform.io/bnjns/metabase/0.9.1/darwin_arm64/terraform-provider-metabase_v0.9.1"]
2025-01-28T18:08:54.601+0200 [DEBUG] provider: plugin started: path=.terraform/providers/registry.terraform.io/bnjns/metabase/0.9.1/darwin_arm64/terraform-provider-metabase_v0.9.1 pid=36819
2025-01-28T18:08:54.601+0200 [DEBUG] provider: waiting for RPC address: plugin=.terraform/providers/registry.terraform.io/bnjns/metabase/0.9.1/darwin_arm64/terraform-provider-metabase_v0.9.1
2025-01-28T18:08:54.610+0200 [INFO]  provider.terraform-provider-metabase_v0.9.1: configuring server automatic mTLS: timestamp="2025-01-28T18:08:54.610+0200"
2025-01-28T18:08:54.616+0200 [DEBUG] provider: using plugin: version=6
2025-01-28T18:08:54.616+0200 [DEBUG] provider.terraform-provider-metabase_v0.9.1: plugin address: address=/var/folders/n0/8075pklx4d3d9dl_jxjlb1r40000gn/T/plugin437385435 network=unix timestamp="2025-01-28T18:08:54.616+0200"
2025-01-28T18:08:54.622+0200 [DEBUG] No provider meta schema returned
2025-01-28T18:08:54.624+0200 [DEBUG] provider.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = error reading from server: EOF"
2025-01-28T18:08:54.624+0200 [INFO]  provider: plugin process exited: plugin=.terraform/providers/registry.terraform.io/bnjns/metabase/0.9.1/darwin_arm64/terraform-provider-metabase_v0.9.1 id=36819
2025-01-28T18:08:54.624+0200 [DEBUG] provider: plugin exited
2025-01-28T18:08:54.624+0200 [DEBUG] ReferenceTransformer: "metabase_database.mongo (expand)" references: []
2025-01-28T18:08:54.624+0200 [DEBUG] ReferenceTransformer: "provider[\"registry.terraform.io/bnjns/metabase\"]" references: []
2025-01-28T18:08:54.624+0200 [DEBUG] Starting graph walk: walkImport
2025-01-28T18:08:54.624+0200 [DEBUG] created provider logger: level=debug
2025-01-28T18:08:54.624+0200 [INFO]  provider: configuring client automatic mTLS
2025-01-28T18:08:54.627+0200 [DEBUG] provider: starting plugin: path=.terraform/providers/registry.terraform.io/bnjns/metabase/0.9.1/darwin_arm64/terraform-provider-metabase_v0.9.1 args=[".terraform/providers/registry.terraform.io/bnjns/metabase/0.9.1/darwin_arm64/terraform-provider-metabase_v0.9.1"]
2025-01-28T18:08:54.628+0200 [DEBUG] provider: plugin started: path=.terraform/providers/registry.terraform.io/bnjns/metabase/0.9.1/darwin_arm64/terraform-provider-metabase_v0.9.1 pid=36820
2025-01-28T18:08:54.628+0200 [DEBUG] provider: waiting for RPC address: plugin=.terraform/providers/registry.terraform.io/bnjns/metabase/0.9.1/darwin_arm64/terraform-provider-metabase_v0.9.1
2025-01-28T18:08:54.638+0200 [INFO]  provider.terraform-provider-metabase_v0.9.1: configuring server automatic mTLS: timestamp="2025-01-28T18:08:54.638+0200"
2025-01-28T18:08:54.643+0200 [DEBUG] provider.terraform-provider-metabase_v0.9.1: plugin address: address=/var/folders/n0/8075pklx4d3d9dl_jxjlb1r40000gn/T/plugin3866794312 network=unix timestamp="2025-01-28T18:08:54.643+0200"
2025-01-28T18:08:54.643+0200 [DEBUG] provider: using plugin: version=6
2025-01-28T18:08:54.649+0200 [DEBUG] ReferenceTransformer: "metabase_database.mongo (import id \"6\")" references: []
metabase_database.mongo: Importing from ID "6"...
metabase_database.mongo: Import prepared!
  Prepared metabase_database for import
metabase_database.mongo: Refreshing state... [name=(REDACTED)]]
2025-01-28T18:08:54.836+0200 [DEBUG] provider.terraform-provider-metabase_v0.9.1: Value switched to prior value due to semantic equality logic: @module=sdk.framework tf_attribute_path=schedules.cache_field_values.type tf_req_id=bf362d4f-174d-c0b6-8a38-75e577ea726f tf_resource_type=metabase_database tf_rpc=ReadResource @caller=github.com/hashicorp/terraform-plugin-framework@v1.9.0/internal/fwschemadata/value_semantic_equality.go:87 tf_provider_addr=registry.terraform.io/bnjns/metabase timestamp="2025-01-28T18:08:54.836+0200"
2025-01-28T18:08:54.836+0200 [DEBUG] provider.terraform-provider-metabase_v0.9.1: Value switched to prior value due to semantic equality logic: tf_req_id=bf362d4f-174d-c0b6-8a38-75e577ea726f tf_resource_type=metabase_database @caller=github.com/hashicorp/terraform-plugin-framework@v1.9.0/internal/fwschemadata/value_semantic_equality.go:87 @module=sdk.framework tf_attribute_path=schedules.cache_field_values.hour tf_provider_addr=registry.terraform.io/bnjns/metabase tf_rpc=ReadResource timestamp="2025-01-28T18:08:54.836+0200"
2025-01-28T18:08:54.836+0200 [DEBUG] provider.terraform-provider-metabase_v0.9.1: Value switched to prior value due to semantic equality logic: @caller=github.com/hashicorp/terraform-plugin-framework@v1.9.0/internal/fwschemadata/value_semantic_equality.go:87 tf_provider_addr=registry.terraform.io/bnjns/metabase tf_resource_type=metabase_database tf_rpc=ReadResource @module=sdk.framework tf_attribute_path=schedules.cache_field_values.minute tf_req_id=bf362d4f-174d-c0b6-8a38-75e577ea726f timestamp="2025-01-28T18:08:54.836+0200"
2025-01-28T18:08:54.836+0200 [DEBUG] provider.terraform-provider-metabase_v0.9.1: Value switched to prior value due to semantic equality logic: @module=sdk.framework tf_attribute_path=schedules.cache_field_values tf_req_id=bf362d4f-174d-c0b6-8a38-75e577ea726f tf_resource_type=metabase_database tf_provider_addr=registry.terraform.io/bnjns/metabase tf_rpc=ReadResource @caller=github.com/hashicorp/terraform-plugin-framework@v1.9.0/internal/fwschemadata/value_semantic_equality.go:87 timestamp="2025-01-28T18:08:54.836+0200"
2025-01-28T18:08:54.836+0200 [DEBUG] provider.terraform-provider-metabase_v0.9.1: Value switched to prior value due to semantic equality logic: tf_rpc=ReadResource @module=sdk.framework tf_attribute_path=schedules.metadata_sync.minute tf_provider_addr=registry.terraform.io/bnjns/metabase tf_req_id=bf362d4f-174d-c0b6-8a38-75e577ea726f tf_resource_type=metabase_database @caller=github.com/hashicorp/terraform-plugin-framework@v1.9.0/internal/fwschemadata/value_semantic_equality.go:87 timestamp="2025-01-28T18:08:54.836+0200"
2025-01-28T18:08:54.836+0200 [DEBUG] provider.terraform-provider-metabase_v0.9.1: Value switched to prior value due to semantic equality logic: @module=sdk.framework tf_provider_addr=registry.terraform.io/bnjns/metabase tf_resource_type=metabase_database tf_rpc=ReadResource tf_attribute_path=schedules.metadata_sync.type tf_req_id=bf362d4f-174d-c0b6-8a38-75e577ea726f @caller=github.com/hashicorp/terraform-plugin-framework@v1.9.0/internal/fwschemadata/value_semantic_equality.go:87 timestamp="2025-01-28T18:08:54.836+0200"
2025-01-28T18:08:54.836+0200 [DEBUG] provider.terraform-provider-metabase_v0.9.1: Value switched to prior value due to semantic equality logic: @module=sdk.framework tf_req_id=bf362d4f-174d-c0b6-8a38-75e577ea726f @caller=github.com/hashicorp/terraform-plugin-framework@v1.9.0/internal/fwschemadata/value_semantic_equality.go:87 tf_resource_type=metabase_database tf_rpc=ReadResource tf_attribute_path=schedules.metadata_sync tf_provider_addr=registry.terraform.io/bnjns/metabase timestamp="2025-01-28T18:08:54.836+0200"
2025-01-28T18:08:54.836+0200 [DEBUG] provider.terraform-provider-metabase_v0.9.1: Value switched to prior value due to semantic equality logic: @module=sdk.framework tf_resource_type=metabase_database tf_attribute_path=schedules tf_provider_addr=registry.terraform.io/bnjns/metabase tf_req_id=bf362d4f-174d-c0b6-8a38-75e577ea726f tf_rpc=ReadResource @caller=github.com/hashicorp/terraform-plugin-framework@v1.9.0/internal/fwschemadata/value_semantic_equality.go:87 timestamp="2025-01-28T18:08:54.836+0200"
2025-01-28T18:08:54.837+0200 [DEBUG] provider.terraform-provider-metabase_v0.9.1: Value switched to prior value due to semantic equality logic: tf_attribute_path=id tf_provider_addr=registry.terraform.io/bnjns/metabase tf_resource_type=metabase_database @module=sdk.framework tf_req_id=bf362d4f-174d-c0b6-8a38-75e577ea726f tf_rpc=ReadResource @caller=github.com/hashicorp/terraform-plugin-framework@v1.9.0/internal/fwschemadata/value_semantic_equality.go:87 timestamp="2025-01-28T18:08:54.836+0200"
2025-01-28T18:08:54.837+0200 [DEBUG] provider.terraform-provider-metabase_v0.9.1: Value switched to prior value due to semantic equality logic: tf_resource_type=metabase_database tf_rpc=ReadResource @module=sdk.framework tf_provider_addr=registry.terraform.io/bnjns/metabase tf_req_id=bf362d4f-174d-c0b6-8a38-75e577ea726f @caller=github.com/hashicorp/terraform-plugin-framework@v1.9.0/internal/fwschemadata/value_semantic_equality.go:87 tf_attribute_path=engine timestamp="2025-01-28T18:08:54.836+0200"
2025-01-28T18:08:54.837+0200 [DEBUG] provider.terraform-provider-metabase_v0.9.1: Value switched to prior value due to semantic equality logic: @module=sdk.framework tf_attribute_path=name tf_provider_addr=registry.terraform.io/bnjns/metabase @caller=github.com/hashicorp/terraform-plugin-framework@v1.9.0/internal/fwschemadata/value_semantic_equality.go:87 tf_req_id=bf362d4f-174d-c0b6-8a38-75e577ea726f tf_resource_type=metabase_database tf_rpc=ReadResource timestamp="2025-01-28T18:08:54.836+0200"
2025-01-28T18:08:54.837+0200 [DEBUG] provider.terraform-provider-metabase_v0.9.1: Value switched to prior value due to semantic equality logic: @caller=github.com/hashicorp/terraform-plugin-framework@v1.9.0/internal/fwschemadata/value_semantic_equality.go:87 @module=sdk.framework tf_attribute_path=features[0] tf_resource_type=metabase_database tf_rpc=ReadResource tf_provider_addr=registry.terraform.io/bnjns/metabase tf_req_id=bf362d4f-174d-c0b6-8a38-75e577ea726f timestamp="2025-01-28T18:08:54.836+0200"
2025-01-28T18:08:54.837+0200 [DEBUG] provider.terraform-provider-metabase_v0.9.1: Value switched to prior value due to semantic equality logic: @caller=github.com/hashicorp/terraform-plugin-framework@v1.9.0/internal/fwschemadata/value_semantic_equality.go:87 @module=sdk.framework tf_rpc=ReadResource tf_attribute_path=features[1] tf_provider_addr=registry.terraform.io/bnjns/metabase tf_req_id=bf362d4f-174d-c0b6-8a38-75e577ea726f tf_resource_type=metabase_database timestamp="2025-01-28T18:08:54.837+0200"
2025-01-28T18:08:54.837+0200 [DEBUG] provider.terraform-provider-metabase_v0.9.1: Value switched to prior value due to semantic equality logic: @caller=github.com/hashicorp/terraform-plugin-framework@v1.9.0/internal/fwschemadata/value_semantic_equality.go:87 tf_attribute_path=features[2] tf_provider_addr=registry.terraform.io/bnjns/metabase tf_resource_type=metabase_database @module=sdk.framework tf_req_id=bf362d4f-174d-c0b6-8a38-75e577ea726f tf_rpc=ReadResource timestamp="2025-01-28T18:08:54.837+0200"
2025-01-28T18:08:54.837+0200 [DEBUG] provider.terraform-provider-metabase_v0.9.1: Value switched to prior value due to semantic equality logic: tf_resource_type=metabase_database tf_rpc=ReadResource @caller=github.com/hashicorp/terraform-plugin-framework@v1.9.0/internal/fwschemadata/value_semantic_equality.go:87 tf_attribute_path=features[3] tf_provider_addr=registry.terraform.io/bnjns/metabase @module=sdk.framework tf_req_id=bf362d4f-174d-c0b6-8a38-75e577ea726f timestamp="2025-01-28T18:08:54.837+0200"
2025-01-28T18:08:54.837+0200 [DEBUG] provider.terraform-provider-metabase_v0.9.1: Value switched to prior value due to semantic equality logic: tf_attribute_path=features[4] tf_provider_addr=registry.terraform.io/bnjns/metabase tf_resource_type=metabase_database tf_rpc=ReadResource @caller=github.com/hashicorp/terraform-plugin-framework@v1.9.0/internal/fwschemadata/value_semantic_equality.go:87 @module=sdk.framework tf_req_id=bf362d4f-174d-c0b6-8a38-75e577ea726f timestamp="2025-01-28T18:08:54.837+0200"
2025-01-28T18:08:54.837+0200 [DEBUG] provider.terraform-provider-metabase_v0.9.1: Value switched to prior value due to semantic equality logic: tf_req_id=bf362d4f-174d-c0b6-8a38-75e577ea726f @module=sdk.framework tf_attribute_path=features[5] tf_provider_addr=registry.terraform.io/bnjns/metabase tf_rpc=ReadResource @caller=github.com/hashicorp/terraform-plugin-framework@v1.9.0/internal/fwschemadata/value_semantic_equality.go:87 tf_resource_type=metabase_database timestamp="2025-01-28T18:08:54.837+0200"
2025-01-28T18:08:54.837+0200 [DEBUG] provider.terraform-provider-metabase_v0.9.1: Value switched to prior value due to semantic equality logic: tf_resource_type=metabase_database tf_rpc=ReadResource @module=sdk.framework tf_req_id=bf362d4f-174d-c0b6-8a38-75e577ea726f tf_attribute_path=features[6] tf_provider_addr=registry.terraform.io/bnjns/metabase @caller=github.com/hashicorp/terraform-plugin-framework@v1.9.0/internal/fwschemadata/value_semantic_equality.go:87 timestamp="2025-01-28T18:08:54.837+0200"
2025-01-28T18:08:54.837+0200 [DEBUG] provider.terraform-provider-metabase_v0.9.1: Value switched to prior value due to semantic equality logic: tf_req_id=bf362d4f-174d-c0b6-8a38-75e577ea726f tf_resource_type=metabase_database tf_attribute_path=features[7] tf_provider_addr=registry.terraform.io/bnjns/metabase tf_rpc=ReadResource @caller=github.com/hashicorp/terraform-plugin-framework@v1.9.0/internal/fwschemadata/value_semantic_equality.go:87 @module=sdk.framework timestamp="2025-01-28T18:08:54.837+0200"
2025-01-28T18:08:54.837+0200 [DEBUG] provider.terraform-provider-metabase_v0.9.1: Value switched to prior value due to semantic equality logic: tf_attribute_path=features[8] tf_provider_addr=registry.terraform.io/bnjns/metabase @caller=github.com/hashicorp/terraform-plugin-framework@v1.9.0/internal/fwschemadata/value_semantic_equality.go:87 @module=sdk.framework tf_req_id=bf362d4f-174d-c0b6-8a38-75e577ea726f tf_resource_type=metabase_database tf_rpc=ReadResource timestamp="2025-01-28T18:08:54.837+0200"
2025-01-28T18:08:54.837+0200 [DEBUG] provider.terraform-provider-metabase_v0.9.1: Value switched to prior value due to semantic equality logic: tf_resource_type=metabase_database tf_rpc=ReadResource @module=sdk.framework tf_provider_addr=registry.terraform.io/bnjns/metabase tf_req_id=bf362d4f-174d-c0b6-8a38-75e577ea726f @caller=github.com/hashicorp/terraform-plugin-framework@v1.9.0/internal/fwschemadata/value_semantic_equality.go:87 tf_attribute_path=features[9] timestamp="2025-01-28T18:08:54.837+0200"
2025-01-28T18:08:54.837+0200 [DEBUG] provider.terraform-provider-metabase_v0.9.1: Value switched to prior value due to semantic equality logic: tf_attribute_path=features[10] tf_provider_addr=registry.terraform.io/bnjns/metabase tf_resource_type=metabase_database @caller=github.com/hashicorp/terraform-plugin-framework@v1.9.0/internal/fwschemadata/value_semantic_equality.go:87 @module=sdk.framework tf_req_id=bf362d4f-174d-c0b6-8a38-75e577ea726f tf_rpc=ReadResource timestamp="2025-01-28T18:08:54.837+0200"
2025-01-28T18:08:54.837+0200 [DEBUG] provider.terraform-provider-metabase_v0.9.1: Value switched to prior value due to semantic equality logic: tf_provider_addr=registry.terraform.io/bnjns/metabase tf_req_id=bf362d4f-174d-c0b6-8a38-75e577ea726f @module=sdk.framework tf_attribute_path=features[11] tf_resource_type=metabase_database tf_rpc=ReadResource @caller=github.com/hashicorp/terraform-plugin-framework@v1.9.0/internal/fwschemadata/value_semantic_equality.go:87 timestamp="2025-01-28T18:08:54.837+0200"
2025-01-28T18:08:54.837+0200 [DEBUG] provider.terraform-provider-metabase_v0.9.1: Value switched to prior value due to semantic equality logic: @module=sdk.framework tf_attribute_path=features[12] tf_provider_addr=registry.terraform.io/bnjns/metabase tf_req_id=bf362d4f-174d-c0b6-8a38-75e577ea726f tf_resource_type=metabase_database tf_rpc=ReadResource @caller=github.com/hashicorp/terraform-plugin-framework@v1.9.0/internal/fwschemadata/value_semantic_equality.go:87 timestamp="2025-01-28T18:08:54.837+0200"
2025-01-28T18:08:54.837+0200 [DEBUG] provider.terraform-provider-metabase_v0.9.1: Value switched to prior value due to semantic equality logic: tf_provider_addr=registry.terraform.io/bnjns/metabase tf_resource_type=metabase_database tf_rpc=ReadResource @caller=github.com/hashicorp/terraform-plugin-framework@v1.9.0/internal/fwschemadata/value_semantic_equality.go:87 tf_attribute_path=features[13] @module=sdk.framework tf_req_id=bf362d4f-174d-c0b6-8a38-75e577ea726f timestamp="2025-01-28T18:08:54.837+0200"
2025-01-28T18:08:54.837+0200 [DEBUG] provider.terraform-provider-metabase_v0.9.1: Value switched to prior value due to semantic equality logic: tf_attribute_path=features[14] tf_provider_addr=registry.terraform.io/bnjns/metabase tf_resource_type=metabase_database tf_rpc=ReadResource @caller=github.com/hashicorp/terraform-plugin-framework@v1.9.0/internal/fwschemadata/value_semantic_equality.go:87 @module=sdk.framework tf_req_id=bf362d4f-174d-c0b6-8a38-75e577ea726f timestamp="2025-01-28T18:08:54.837+0200"
2025-01-28T18:08:54.837+0200 [DEBUG] provider.terraform-provider-metabase_v0.9.1: Value switched to prior value due to semantic equality logic: @module=sdk.framework tf_attribute_path=features[15] tf_resource_type=metabase_database @caller=github.com/hashicorp/terraform-plugin-framework@v1.9.0/internal/fwschemadata/value_semantic_equality.go:87 tf_provider_addr=registry.terraform.io/bnjns/metabase tf_req_id=bf362d4f-174d-c0b6-8a38-75e577ea726f tf_rpc=ReadResource timestamp="2025-01-28T18:08:54.837+0200"
2025-01-28T18:08:54.837+0200 [DEBUG] provider.terraform-provider-metabase_v0.9.1: Value switched to prior value due to semantic equality logic: tf_req_id=bf362d4f-174d-c0b6-8a38-75e577ea726f tf_resource_type=metabase_database @caller=github.com/hashicorp/terraform-plugin-framework@v1.9.0/internal/fwschemadata/value_semantic_equality.go:87 @module=sdk.framework tf_attribute_path=features[16] tf_provider_addr=registry.terraform.io/bnjns/metabase tf_rpc=ReadResource timestamp="2025-01-28T18:08:54.837+0200"
2025-01-28T18:08:54.837+0200 [DEBUG] provider.terraform-provider-metabase_v0.9.1: Value switched to prior value due to semantic equality logic: tf_resource_type=metabase_database @module=sdk.framework tf_attribute_path=features[17] tf_provider_addr=registry.terraform.io/bnjns/metabase tf_req_id=bf362d4f-174d-c0b6-8a38-75e577ea726f tf_rpc=ReadResource @caller=github.com/hashicorp/terraform-plugin-framework@v1.9.0/internal/fwschemadata/value_semantic_equality.go:87 timestamp="2025-01-28T18:08:54.837+0200"
2025-01-28T18:08:54.837+0200 [DEBUG] provider.terraform-provider-metabase_v0.9.1: Value switched to prior value due to semantic equality logic: tf_rpc=ReadResource @module=sdk.framework tf_attribute_path=features[18] tf_req_id=bf362d4f-174d-c0b6-8a38-75e577ea726f tf_resource_type=metabase_database @caller=github.com/hashicorp/terraform-plugin-framework@v1.9.0/internal/fwschemadata/value_semantic_equality.go:87 tf_provider_addr=registry.terraform.io/bnjns/metabase timestamp="2025-01-28T18:08:54.837+0200"
2025-01-28T18:08:54.837+0200 [DEBUG] provider.terraform-provider-metabase_v0.9.1: Value switched to prior value due to semantic equality logic: @module=sdk.framework tf_attribute_path=features[19] tf_provider_addr=registry.terraform.io/bnjns/metabase tf_req_id=bf362d4f-174d-c0b6-8a38-75e577ea726f tf_rpc=ReadResource @caller=github.com/hashicorp/terraform-plugin-framework@v1.9.0/internal/fwschemadata/value_semantic_equality.go:87 tf_resource_type=metabase_database timestamp="2025-01-28T18:08:54.837+0200"
2025-01-28T18:08:54.837+0200 [DEBUG] provider.terraform-provider-metabase_v0.9.1: Value switched to prior value due to semantic equality logic: @module=sdk.framework tf_provider_addr=registry.terraform.io/bnjns/metabase tf_rpc=ReadResource tf_attribute_path=features[20] tf_req_id=bf362d4f-174d-c0b6-8a38-75e577ea726f tf_resource_type=metabase_database @caller=github.com/hashicorp/terraform-plugin-framework@v1.9.0/internal/fwschemadata/value_semantic_equality.go:87 timestamp="2025-01-28T18:08:54.837+0200"
2025-01-28T18:08:54.837+0200 [DEBUG] provider.terraform-provider-metabase_v0.9.1: Value switched to prior value due to semantic equality logic: tf_req_id=bf362d4f-174d-c0b6-8a38-75e577ea726f tf_resource_type=metabase_database tf_provider_addr=registry.terraform.io/bnjns/metabase tf_rpc=ReadResource @caller=github.com/hashicorp/terraform-plugin-framework@v1.9.0/internal/fwschemadata/value_semantic_equality.go:87 @module=sdk.framework tf_attribute_path=features timestamp="2025-01-28T18:08:54.837+0200"
2025-01-28T18:08:54.838+0200 [ERROR] vertex "import metabase_database.mongo result" error: Provider produced invalid object
2025-01-28T18:08:54.838+0200 [ERROR] vertex "metabase_database.mongo (import id \"6\")" error: Provider produced invalid object
2025-01-28T18:08:54.838+0200 [ERROR] vertex "metabase_database.mongo (expand)" error: Provider produced invalid object
╷
│ Error: Provider produced invalid object
│ 
│ Provider "registry.terraform.io/bnjns/metabase" planned an invalid value for metabase_database.mongo during refresh: The returned state contains unknown values.
│ 
│ This is a bug in the provider, which should be reported in the provider's own issue tracker.
╵

2025-01-28T18:08:54.839+0200 [DEBUG] provider.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = error reading from server: EOF"
2025-01-28T18:08:54.840+0200 [INFO]  provider: plugin process exited: plugin=.terraform/providers/registry.terraform.io/bnjns/metabase/0.9.1/darwin_arm64/terraform-provider-metabase_v0.9.1 id=36820
2025-01-28T18:08:54.840+0200 [DEBUG] provider: plugin exited

```
</details>

### Panic Output

<!--- If Terraform produced a panic, please provide the output of the `crash.log`. --->

### Expected Behavior

<!--- What should have happened? --->
Should have imported the resource

### Actual Behavior

<!--- What actually happened? --->
```
Error: Provider produced invalid object
```

### Steps to Reproduce

<!--- Please list the steps required to reproduce the issue. --->

1. Manually create a MongoDB data source in Metabase
2. Write the resource definition in TF
3. Try to import the data source to TF state

### References

<!---
Information about referencing Github Issues: https://help.github.com/articles/basic-writing-and-formatting-syntax/#referencing-issues-and-pull-requests

Are there any other GitHub issues (open or closed) or pull requests that should be linked here? API documentation? For example:
--->

* #0000

--- a/.github/workflows/docs.yml
+++ b/.github/workflows/docs.yml
@@ -17,6 +17,7 @@ jobs:
     runs-on: ubuntu-latest
     steps:
       - uses: actions/checkout@v4
+      - uses: hashicorp/setup-terraform@v3
       - uses: actions/setup-go@v5
         with:
           go-version-file: 'go.mod'

--- a/docker-compose.yml
+++ b/docker-compose.yml
@@ -36,3 +36,18 @@ services:
       timeout: 5s
       retries: 5
     restart: unless-stopped
+
+  mongo:
+    image: mongo
+    container_name: mb-mongo
+    networks:
+      - main
+    environment:
+      MONGO_INITDB_ROOT_USERNAME: mongo
+      MONGO_INITDB_ROOT_PASSWORD: mongo
+    healthcheck:
+      test: ["CMD-SHELL", "[ $(mongosh \"mongodb://mongo:mongo@localhost:27017/\" --eval 'db.adminCommand(\"ping\").ok') -eq 1 ]"]
+      interval: 10s
+      timeout: 5s
+      retries: 5
+    restart: unless-stopped

--- a/internal/provider/database_data_source.go
+++ b/internal/provider/database_data_source.go
@@ -74,7 +74,7 @@ func mapDatabaseToDataSource(ctx context.Context, db *database.Database, target 
 	target.Schedules = schedules
 	diags.Append(scheduleDiags...)
 
-	details, _, detailsDiags := buildDetails(db)
+	details, _, detailsDiags := buildDatabaseDetails(db)
 	target.Details = details
 	diags.Append(detailsDiags...)
 

--- a/internal/provider/database_data_source.go
+++ b/internal/provider/database_data_source.go
@@ -2,19 +2,13 @@ package provider
 
 import (
 	"context"
-	"encoding/json"
 	"fmt"
 	"github.com/bnjns/metabase-sdk-go/service/database"
 	"github.com/hashicorp/terraform-plugin-framework/datasource"
 	"github.com/hashicorp/terraform-plugin-framework/diag"
 	"github.com/hashicorp/terraform-plugin-framework/types"
-	"golang.org/x/exp/slices"
-	"regexp"
 	"terraform-provider-metabase/internal/schema"
 )
-
-var sensitiveDatabaseDetails = []string{"password", "service-account-json"}
-var redactedPattern = regexp.MustCompile(`^\*\*.+\*\*$`)
 
 // Ensure provider defined types fully satisfy framework interfaces
 var _ datasource.DataSource = &DatabaseDataSource{}
@@ -80,37 +74,9 @@ func mapDatabaseToDataSource(ctx context.Context, db *database.Database, target 
 	target.Schedules = schedules
 	diags.Append(scheduleDiags...)
 
-	details := make(map[string]interface{})
-	for k, v := range *db.Details {
-		if shouldIncludeDatabaseDetail(k, v) {
-			details[k] = v
-		}
-	}
-	detailsBytes, detailsErr := json.Marshal(details)
-	if detailsErr != nil {
-		diags.AddError(
-			fmt.Sprintf("Error fetching database with ID: %d", db.Id),
-			fmt.Sprintf("An error occurred when serialising the details: %s", detailsErr.Error()),
-		)
-	}
-	target.Details = types.StringValue(string(detailsBytes))
+	details, _, detailsDiags := buildDetails(db)
+	target.Details = details
+	diags.Append(detailsDiags...)
 
 	return diags
-}
-
-func shouldIncludeDatabaseDetail(key string, value interface{}) bool {
-	if slices.Contains(sensitiveDatabaseDetails, key) {
-		return false
-	}
-
-	valueStr, isString := value.(string)
-	if !isString {
-		return true
-	}
-
-	if redactedPattern.MatchString(valueStr) {
-		return false
-	}
-
-	return true
 }

--- a/internal/provider/database_data_source_test.go
+++ b/internal/provider/database_data_source_test.go
@@ -2,37 +2,8 @@ package provider
 
 import (
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
-	"github.com/stretchr/testify/assert"
 	"testing"
 )
-
-func TestShouldIncludeDatabaseDetails(t *testing.T) {
-	t.Parallel()
-
-	t.Run("detail with sensitive key should not be included", func(t *testing.T) {
-		result := shouldIncludeDatabaseDetail("password", "example")
-
-		assert.False(t, result)
-	})
-
-	t.Run("detail with a non-string value should be included", func(t *testing.T) {
-		result := shouldIncludeDatabaseDetail("port", 5432)
-
-		assert.True(t, result)
-	})
-
-	t.Run("detail with a redacted string value should not be included", func(t *testing.T) {
-		result := shouldIncludeDatabaseDetail("field", "**MetabasePass**")
-
-		assert.False(t, result)
-	})
-
-	t.Run("detail with a non-redacted string value should be included", func(t *testing.T) {
-		result := shouldIncludeDatabaseDetail("host", "localhost")
-
-		assert.True(t, result)
-	})
-}
 
 func TestAccDatabaseDataSource_Basic(t *testing.T) {
 	resource.Test(t, resource.TestCase{

--- a/internal/provider/database_resource.go
+++ b/internal/provider/database_resource.go
@@ -11,8 +11,9 @@ import (
 	"github.com/hashicorp/terraform-plugin-framework/resource"
 	"github.com/hashicorp/terraform-plugin-framework/types"
 	"github.com/hashicorp/terraform-plugin-framework/types/basetypes"
+	"golang.org/x/exp/slices"
+	"regexp"
 	"strconv"
-	"strings"
 	"terraform-provider-metabase/internal/schema"
 	"terraform-provider-metabase/internal/transforms"
 	"terraform-provider-metabase/internal/utils"
@@ -41,7 +42,8 @@ var (
 	errMissingUser       = errors.New("you must provide the auth username in the 'user' property")
 	errMissingPassword   = errors.New("you must provide the auth password in the 'password' property")
 )
-var databaseSensitiveProperties = []string{"password"}
+var sensitiveDatabaseDetails = []string{"password", "service-account-json"}
+var redactedPattern = regexp.MustCompile(`^\*\*.+\*\*$`)
 
 type DatabaseResource struct {
 	provider *MetabaseProvider
@@ -294,6 +296,23 @@ func mapDatabaseToState(ctx context.Context, db *database.Database, target *Data
 	return diags
 }
 
+func isSensitiveDatabaseDetail(key string, value interface{}) bool {
+	if slices.Contains(sensitiveDatabaseDetails, key) {
+		return true
+	}
+
+	valueStr, isString := value.(string)
+	if !isString {
+		return false
+	}
+
+	if redactedPattern.MatchString(valueStr) {
+		return true
+	}
+
+	return false
+}
+
 func buildDetails(db *database.Database) (types.String, types.String, diag.Diagnostics) {
 	var diags diag.Diagnostics
 
@@ -304,14 +323,9 @@ func buildDetails(db *database.Database) (types.String, types.String, diag.Diagn
 	details := make(map[string]any)
 	detailsSecure := make(map[string]any)
 	for k, v := range *db.Details {
-		switch v.(type) {
-		case string:
-			if strings.HasPrefix(v.(string), "**") && strings.HasSuffix(v.(string), "**") {
-				detailsSecure[k] = v
-			} else {
-				details[k] = v
-			}
-		default:
+		if isSensitiveDatabaseDetail(k, v) {
+			detailsSecure[k] = v
+		} else {
 			details[k] = v
 		}
 	}

--- a/internal/provider/database_resource.go
+++ b/internal/provider/database_resource.go
@@ -2,6 +2,7 @@ package provider
 
 import (
 	"context"
+	"encoding/json"
 	"errors"
 	"fmt"
 	"github.com/bnjns/metabase-sdk-go/service/database"
@@ -11,6 +12,7 @@ import (
 	"github.com/hashicorp/terraform-plugin-framework/types"
 	"github.com/hashicorp/terraform-plugin-framework/types/basetypes"
 	"strconv"
+	"strings"
 	"terraform-provider-metabase/internal/schema"
 	"terraform-provider-metabase/internal/transforms"
 	"terraform-provider-metabase/internal/utils"
@@ -260,10 +262,14 @@ func (d *DatabaseResource) fetchDatabaseState(ctx context.Context, databaseId in
 	var state DatabaseModel
 	diags := mapDatabaseToState(ctx, db, &state)
 
-	// Override both details and details_secure
-	// The API can return additional keys in details, and anything in details_secure is redacted so the response is useless
-	state.Details = plan.Details
-	state.DetailsSecure = plan.DetailsSecure
+	// Override both details and details_secure if they're set in the plan as the API can return additional keys in
+	// details and this produces an inconsistent result
+	if !plan.Details.IsUnknown() {
+		state.Details = plan.Details
+	}
+	if !plan.DetailsSecure.IsUnknown() {
+		state.DetailsSecure = plan.DetailsSecure
+	}
 
 	return state, diags
 }
@@ -275,13 +281,52 @@ func mapDatabaseToState(ctx context.Context, db *database.Database, target *Data
 	target.Engine = types.StringValue(string(db.Engine))
 	target.Name = types.StringValue(db.Name)
 	target.Features, _ = types.ListValueFrom(ctx, types.StringType, db.Features)
-	target.Details = types.StringUnknown()
-	target.DetailsSecure = types.StringUnknown()
 
 	schedules, scheduleDiags := buildSchedules(db)
 	target.Schedules = schedules
 	diags.Append(scheduleDiags...)
+
+	details, detailsSecure, detailsDiags := buildDetails(db)
+	target.Details = details
+	target.DetailsSecure = detailsSecure
+	diags.Append(detailsDiags...)
+
 	return diags
+}
+
+func buildDetails(db *database.Database) (types.String, types.String, diag.Diagnostics) {
+	var diags diag.Diagnostics
+
+	if db.Details == nil {
+		return types.StringNull(), types.StringNull(), diags
+	}
+
+	details := make(map[string]any)
+	detailsSecure := make(map[string]any)
+	for k, v := range *db.Details {
+		switch v.(type) {
+		case string:
+			if strings.HasPrefix(v.(string), "**") && strings.HasSuffix(v.(string), "**") {
+				detailsSecure[k] = v
+			} else {
+				details[k] = v
+			}
+		default:
+			details[k] = v
+		}
+	}
+
+	detailsStr, err := json.Marshal(details)
+	if err != nil {
+		diags = append(diags, diag.NewErrorDiagnostic(fmt.Sprintf("Error parsing details for database %d", db.Id), err.Error()))
+	}
+
+	detailsSecureStr, err := json.Marshal(detailsSecure)
+	if err != nil {
+		diags = append(diags, diag.NewErrorDiagnostic(fmt.Sprintf("Error parsing details_secure for database %d", db.Id), err.Error()))
+	}
+
+	return types.StringValue(string(detailsStr)), types.StringValue(string(detailsSecureStr)), diags
 }
 
 func buildScheduleSettings(settings *database.ScheduleSettings) basetypes.ObjectValue {

--- a/internal/provider/database_resource.go
+++ b/internal/provider/database_resource.go
@@ -288,7 +288,7 @@ func mapDatabaseToState(ctx context.Context, db *database.Database, target *Data
 	target.Schedules = schedules
 	diags.Append(scheduleDiags...)
 
-	details, detailsSecure, detailsDiags := buildDetails(db)
+	details, detailsSecure, detailsDiags := buildDatabaseDetails(db)
 	target.Details = details
 	target.DetailsSecure = detailsSecure
 	diags.Append(detailsDiags...)
@@ -313,7 +313,7 @@ func isSensitiveDatabaseDetail(key string, value interface{}) bool {
 	return false
 }
 
-func buildDetails(db *database.Database) (types.String, types.String, diag.Diagnostics) {
+func buildDatabaseDetails(db *database.Database) (types.String, types.String, diag.Diagnostics) {
 	var diags diag.Diagnostics
 
 	if db.Details == nil {

--- a/internal/provider/database_resource_test.go
+++ b/internal/provider/database_resource_test.go
@@ -125,6 +125,47 @@ func TestBuildSchedules(t *testing.T) {
 	})
 }
 
+func TestBuildDatabaseDetails(t *testing.T) {
+	t.Parallel()
+
+	t.Run("a database with no details should return null", func(t *testing.T) {
+		db := database.Database{
+			Details: nil,
+		}
+
+		details, detailsSecure, diags := buildDatabaseDetails(&db)
+		assert.Zero(t, len(diags))
+		assert.True(t, details.IsNull())
+		assert.True(t, detailsSecure.IsNull())
+	})
+
+	t.Run("a database with an empty details map should be parsed as empty json objects", func(t *testing.T) {
+		db := database.Database{
+			Details: &database.Details{},
+		}
+
+		details, detailsSecure, diags := buildDatabaseDetails(&db)
+		assert.Zero(t, len(diags))
+		assert.Equal(t, `{}`, details.ValueString())
+		assert.Equal(t, `{}`, detailsSecure.ValueString())
+	})
+
+	t.Run("a database with details should have the sensitive details separated", func(t *testing.T) {
+		db := database.Database{
+			Details: &database.Details{
+				"password": "password",
+				"name":     "database name",
+				"redacted": "**MetabasePass**",
+			},
+		}
+
+		details, detailsSecure, diags := buildDatabaseDetails(&db)
+		assert.Zero(t, len(diags))
+		assert.Equal(t, `{"name":"database name"}`, details.ValueString())
+		assert.Equal(t, `{"password":"password","redacted":"**MetabasePass**"}`, detailsSecure.ValueString())
+	})
+}
+
 func TestAccDatabaseResource_PostgreSQL(t *testing.T) {
 	resource.Test(t, resource.TestCase{
 		ProtoV6ProviderFactories: testAccProtoV6ProviderFactories,

--- a/internal/provider/database_resource_test.go
+++ b/internal/provider/database_resource_test.go
@@ -9,6 +9,34 @@ import (
 
 var ignoredDatabaseImportAttributes = []string{"details", "details_secure"}
 
+func TestIsSensitiveDatabaseDetail(t *testing.T) {
+	t.Parallel()
+
+	t.Run("detail with sensitive key should be sensitive", func(t *testing.T) {
+		result := isSensitiveDatabaseDetail("password", "example")
+
+		assert.True(t, result)
+	})
+
+	t.Run("detail with a non-string value should not be sensitive", func(t *testing.T) {
+		result := isSensitiveDatabaseDetail("port", 5432)
+
+		assert.False(t, result)
+	})
+
+	t.Run("detail with a redacted string value should be sensitive", func(t *testing.T) {
+		result := isSensitiveDatabaseDetail("field", "**MetabasePass**")
+
+		assert.True(t, result)
+	})
+
+	t.Run("detail with a non-redacted string value should not be sensitive", func(t *testing.T) {
+		result := isSensitiveDatabaseDetail("host", "localhost")
+
+		assert.False(t, result)
+	})
+}
+
 func TestCheckDatabaseDetails(t *testing.T) {
 	t.Parallel()
 

